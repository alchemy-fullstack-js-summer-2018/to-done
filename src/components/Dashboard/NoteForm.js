import React, { Component } from 'react';
import PropTypes from 'prop-types';

class NoteForm extends Component {

  state = {
    editing: false,
    key: null,
    title: '',
    notes: '',
    completed: false
  };

  static propTypes = {
    note: PropTypes.object,
    onComplete: PropTypes.func.isRequired,
    onCancel: PropTypes.func
  };

  componentDidMount() {
    const { note } = this.props;
    if(!note) return;

    this.setState(note);
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { key, title, notes, completed } = this.state;
    const note = { title, notes, completed };
    if(key) note.key = key;

    this.props.onComplete(note)
      .then(() => {
        if(!key) return;
        this.setState({ title: '', notes: '', completed: false });
      });
  };

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value });
  };

  render() {
    const { key, title, notes, completed } = this.state;
    const { onCancel } = this.props;

    return (
      <div>
        <h3>Add a note:</h3>
        <form onSubmit={this.handleSubmit}>
          <label>Title:<input type="text" name="title" value={title} onChange={this.handleChange}></input></label>
          <label>Notes:<textarea rows="4" cols="50" name="notes" value={notes} onChange={this.handleChange}></textarea></label>
          <label>Completed:<input type="checkbox" name="completed" value={completed} onChange={this.handleChange}></input></label>
          <p>
            <button type="submit">Submit</button>
            {key && <button type="button" onClick={onCancel}>Cancel</button>}
          </p>
        </form>
      </div>
    );
  }
}

export default NoteForm;