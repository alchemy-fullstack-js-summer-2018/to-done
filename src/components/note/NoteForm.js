import React, { Component } from 'react';
import PropTypes from 'prop-types';

class NoteForm extends Component {

  state = {
    editing: false,
    key: null,
    name: '',
    content: ''
  };

  static propTypes = {
    note: PropTypes.object,
    onComplete: PropTypes.func.isRequired,
    onCancel: PropTypes.func
  };

  componentDidMount() {
    const { note } = this.props;
    if(!note) return;

    this.setState(note);
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const { name, content, key } = this.state;
    const note = { name, content };
    if(key) note.key = key;

    this.props.onComplete(note)
      .then(() => {
        if(!key) return;
        this.setState({ name: '', content: '' });
      });
  };

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value });
  };

  render() {
    const { key, name, content } = this.state;
    const { onCancel } = this.props;

    return (
      <form onSubmit={this.handleSubmit}>
        <InputControl name="name" value={ name } onChange={this.handleChange}/>
        <InputControl name="content" value={ content } onChange={this.handleChange}/>
        <p>
          <button type="submit">{ key ? 'Update' : 'Add' }</button>
          {key && <button type="button" onClick={onCancel}>Cancel</button>}
        </p>
      </form>
    );
  }
}

const InputControl = (props) => (
  <p>
    <label>
      {props.name}:
      <input {...props} required/>
    </label>
  </p>
);

export default NoteForm;