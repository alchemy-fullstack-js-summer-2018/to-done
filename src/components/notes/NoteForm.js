import React, { Component } from 'react';
import PropTypes from 'prop-types';

class NoteForm extends Component {
  state = {
    editing: false,
    key: null,
    completed: false,
    content: '',
    title: ''
  };

  static propTypes = {
    note: PropTypes.object,
    onComplete: PropTypes.func.isRequired,
    onCancel: PropTypes.func
  };

  componentDidMount() {
    const { note } = this.props;
    if(!note) return;

    this.setState(note);
  }

  handleSubmit = event => {
    event.preventDefault();
    const { title, content, completed, key } = this.state;
    const note = { title, content, completed };
    if(key) note.key = key;

    this.props.onComplete(note)
      .then(() => {
        if(!key) return;
        this.setState({ title: '', content: '', completed: false });
      });
  };

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value });
  };

  render() {
    const { key, title, content, completed } = this.state;
    const { onCancel } = this.props;

    return (
      <form onSubmit={this.handleSubmit}>
        <InputControl name="title" value={title} onChange={this.handleChange}/>
        <InputControl name="content" value={content} onChange={this.handleChange}/>
        <label>Completed? </label>
        <input name="completed" value={completed} type="checkbox" onChange={this.handleChange}/>
        <p>
          <button type="submit">{ key ? 'Update' : 'Add' }</button>
          {key && <button type="button" onClick={onCancel}>Cancel</button>}
        </p>
      </form>
    );
  }
}

const InputControl = props => (
  <p>
    <label>
      {props.name}:
      <input {...props} required />
    </label>
  </p>
);
 
export default NoteForm;