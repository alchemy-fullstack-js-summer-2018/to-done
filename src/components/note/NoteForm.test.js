import NoteForm from './NoteForm';
import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';

const Enzyme = require('enzyme');
const Adapter = require('enzyme-adapter-react-16');

Enzyme.configure({ adapter: new Adapter() });

describe('Note Form', () => {
  
  it('renders add if no note prop', () => {
    const handleComplete = jest.fn();
    const promise = Promise.resolve();
    handleComplete.mockReturnValueOnce(promise);

    const wrapper = mount(<NoteForm onComplete={handleComplete}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();

    const note = {
      name: 'wash',
      content: 'hair'
    };

    wrapper.find('input[name="name"]').simulate('change', {
      target: { 
        name: 'name',
        value: note.name
      }
    });
    
    wrapper.find('input[name="content"]').simulate('change', {
      target: { 
        name: 'content',
        value: note.content 
      }
    });

    wrapper.find('button').simulate('submit');

    const calls = handleComplete.mock.calls;
    expect(calls.length).toBe(1); 
    expect(calls[0][0]).toEqual(note);

    expect(toJSON(wrapper)).toMatchSnapshot();

    return promise
      .then(() => {
        wrapper.update();
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
  });
  
  it('renders edit if note prop', () => {
    const handleComplete = jest.fn();
    const promise = Promise.resolve();
    handleComplete.mockReturnValueOnce(promise);
    const handleCancel = jest.fn();

    const note = { key: 'abc', name: 'tonight', content: 'cook dinner' };

    const wrapper = mount(<NoteForm 
      onComplete={handleComplete}
      onCancel={handleCancel}
      note={note}
    />);

    expect(toJSON(wrapper)).toMatchSnapshot();

    wrapper.find('input[name="name"]').simulate('change', {
      target: { 
        name: 'name',
        value: 'wash' 
      }
    });

    wrapper.find('button[type="submit"]').simulate('submit');

    const calls = handleComplete.mock.calls;
    expect(calls.length).toBe(1); 
    expect(calls[0][0]).toEqual({
      ...note,
      name: 'wash'
    });

    return promise
      .then(() => {
        wrapper.update();
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
  });
});
