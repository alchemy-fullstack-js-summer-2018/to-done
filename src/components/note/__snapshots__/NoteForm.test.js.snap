// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note Form renders add if no note prop 1`] = `
<NoteForm
  onComplete={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="content"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          content
          :
          <input
            name="content"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</NoteForm>
`;

exports[`Note Form renders add if no note prop 2`] = `
<NoteForm
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "content": "hair",
            "name": "wash",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Promise {},
        },
      ],
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value="wash"
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value="wash"
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="content"
      onChange={[Function]}
      value="hair"
    >
      <p>
        <label>
          content
          :
          <input
            name="content"
            onChange={[Function]}
            required={true}
            value="hair"
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</NoteForm>
`;

exports[`Note Form renders add if no note prop 3`] = `
<NoteForm
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "content": "hair",
            "name": "wash",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Promise {},
        },
      ],
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value="wash"
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value="wash"
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="content"
      onChange={[Function]}
      value="hair"
    >
      <p>
        <label>
          content
          :
          <input
            name="content"
            onChange={[Function]}
            required={true}
            value="hair"
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</NoteForm>
`;

exports[`Note Form renders edit if note prop 1`] = `
<NoteForm
  note={
    Object {
      "content": "cook dinner",
      "key": "abc",
      "name": "tonight",
    }
  }
  onCancel={[MockFunction]}
  onComplete={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value="tonight"
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value="tonight"
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="content"
      onChange={[Function]}
      value="cook dinner"
    >
      <p>
        <label>
          content
          :
          <input
            name="content"
            onChange={[Function]}
            required={true}
            value="cook dinner"
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Update
      </button>
      <button
        onClick={[MockFunction]}
        type="button"
      >
        Cancel
      </button>
    </p>
  </form>
</NoteForm>
`;

exports[`Note Form renders edit if note prop 2`] = `
<NoteForm
  note={
    Object {
      "content": "cook dinner",
      "key": "abc",
      "name": "tonight",
    }
  }
  onCancel={[MockFunction]}
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "content": "cook dinner",
            "key": "abc",
            "name": "wash",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Promise {},
        },
      ],
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="content"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          content
          :
          <input
            name="content"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Update
      </button>
      <button
        onClick={[MockFunction]}
        type="button"
      >
        Cancel
      </button>
    </p>
  </form>
</NoteForm>
`;
